From e3fe82eba47c78758a21d8413c52063fc105550e Mon Sep 17 00:00:00 2001
From: Colin Watson <cjwatson@debian.org>
Date: Sun, 28 Jun 2026 15:52:39 +0100
Subject: [PATCH] Fix use of pytest.warns(match=...)

When called with a callable as a positional argument after the expected
warning, `pytest.warns` either ignores the `match` keyword argument
(pytest < 9.1.0) or passes it on to the provided callable (pytest >=
9.1.0).  In the former case it was merely ineffective, masking an
apparent bug (#13356), while in the latter case it results in
`TypeError: ... got an unexpected keyword argument 'match'`.
---
 tests/test_deprecated_fields.py | 33 +++++++++++++++++++++++----------
 1 file changed, 23 insertions(+), 10 deletions(-)

diff --git a/tests/test_deprecated_fields.py b/tests/test_deprecated_fields.py
index b17f037ede2..a677a8301bc 100644
--- a/tests/test_deprecated_fields.py
+++ b/tests/test_deprecated_fields.py
@@ -25,7 +25,8 @@ class Model(BaseModel):
 
     instance = Model(a=1, b=1, c=1)
 
-    pytest.warns(DeprecationWarning, lambda: instance.a, match='^$')
+    with pytest.warns(DeprecationWarning, match='^$'):
+        instance.a
 
     with pytest.warns(DeprecationWarning, match='^This is deprecated$'):
         b = instance.b
@@ -52,9 +53,12 @@ class Model(BaseModel):
 
     instance = Model(a=1)
 
-    pytest.warns(DeprecationWarning, lambda: instance.a, match='^$')
-    pytest.warns(DeprecationWarning, lambda: instance.b, match='^This is deprecated$')
-    pytest.warns(DeprecationWarning, lambda: instance.c, match='^This is deprecated$')
+    with pytest.warns(DeprecationWarning, match='^$'):
+        instance.a
+    with pytest.warns(DeprecationWarning, match='^This is deprecated$'):
+        instance.b
+    with pytest.warns(DeprecationWarning, match='^This is deprecated$'):
+        instance.c
 
 
 def test_deprecated_fields_field_validator():
@@ -147,10 +151,17 @@ def p5(self) -> int:
 
     instance = Model()
 
-    pytest.warns(DeprecationWarning, lambda: instance.p1, match='^This is deprecated$')
-    pytest.warns(DeprecationWarning, lambda: instance.p2, match='^This is deprecated$')
-    pytest.warns(DeprecationWarning, lambda: instance.p4, match='^This is deprecated$')
-    pytest.warns(DeprecationWarning, lambda: instance.p5, match='^This is deprecated$')
+    with pytest.warns(DeprecationWarning, match='^This is deprecated$'):
+        instance.p1
+    # TODO: This should be match='^This is deprecated$' instead, but the
+    # overriding mechanism has apparently never worked:
+    # https://github.com/pydantic/pydantic/issues/13356
+    with pytest.warns(DeprecationWarning, match=r'^This is deprecated \(this message is overridden\)$'):
+        instance.p2
+    with pytest.warns(DeprecationWarning, match='^This is deprecated$'):
+        instance.p4
+    with pytest.warns(DeprecationWarning, match='^This is deprecated$'):
+        instance.p5
 
     with pytest.warns(DeprecationWarning, match='^$'):
         p3 = instance.p3
@@ -219,7 +230,8 @@ class Model(BaseModel):
 
     instance = Model(a=1, b=1)
 
-    pytest.warns(DeprecationWarning, lambda: instance.a, match='deprecated')
+    with pytest.warns(DeprecationWarning, match='deprecated'):
+        instance.a
 
 
 def test_computed_field_deprecated_class_access() -> None:
@@ -257,7 +269,8 @@ class Model(BaseModel):
 
     m = Model()
 
-    pytest.warns(DeprecationWarning, lambda: m.a, match='test')
+    with pytest.warns(DeprecationWarning, match='test'):
+        m.a
 
 
 def test_deprecated_field_with_assignment() -> None:
